src/std/io/bio/inline.ss

```diff
@@ -3,49 +3,61 @@
 ;;; Buffered IO inline operations
 (import :std/sugar
         :std/interface
+        ../interface
         ./types
         ./input
         ./output)
 (export #t)
 
+(defrule (is-input-buffer? reader)
+  (input-buffer? (&interface-instance-object reader)))
+(defrule (is-output-buffer? writer)
+  (output-buffer? (&interface-instance-object writer)))
+
 (defrule (&BufferedReader-read-u8-inline reader)
-  (let ()
-    (declare (not interrupts-enabled))
-    (let (bio (&interface-instance-object reader))
-      (let ((rlo (&input-buffer-rlo bio))
-            (rhi (&input-buffer-rhi bio))
-            (buf (&input-buffer-buf bio)))
-        (if (fx< rlo rhi)
-          (let ((u8 (u8vector-ref buf rlo))
-                (rlo+1 (fx+ rlo 1)))
-            (bio-input-advance! bio rlo+1 rhi)
-            u8)
-          ;; empty buffer, fall back to the method
-          (bio-read-u8 bio))))))
+  (if (is-input-buffer? reader)
+    (let ()
+      (declare (not interrupts-enabled))
+      (let (bio (&interface-instance-object reader))
+        (let ((rlo (&input-buffer-rlo bio))
+              (rhi (&input-buffer-rhi bio))
+              (buf (&input-buffer-buf bio)))
+          (if (fx< rlo rhi)
+            (let ((u8 (u8vector-ref buf rlo))
+                  (rlo+1 (fx+ rlo 1)))
+              (bio-input-advance! bio rlo+1 rhi)
+              u8)
+            ;; empty buffer, fall back to the method
+            (bio-read-u8 bio)))))
+    (&BufferedReader-read-u8 reader)))
 
 (defrule (&BufferedReader-peek-u8-inline reader)
-  (let ()
-    (declare (not interrupts-enabled))
-    (let (bio (&interface-instance-object reader))
-      (let ((rlo (&input-buffer-rlo bio))
-            (rhi (&input-buffer-rhi bio))
-            (buf (&input-buffer-buf bio)))
-        (if (fx< rlo rhi)
-          (u8vector-ref buf rlo)
-          ;; empty buffer, fall back to the method
-          (bio-peek-u8 bio))))))
+  (if (is-input-buffer? reader)
+    (let ()
+      (declare (not interrupts-enabled))
+      (let (bio (&interface-instance-object reader))
+        (let ((rlo (&input-buffer-rlo bio))
+              (rhi (&input-buffer-rhi bio))
+              (buf (&input-buffer-buf bio)))
+          (if (fx< rlo rhi)
+            (u8vector-ref buf rlo)
+            ;; empty buffer, fall back to the method
+            (bio-peek-u8 bio)))))
+    (&BufferedReader-peek-u8 reader)))
 
 (defrule (&BufferedWriter-write-u8-inline writer u8)
```

Comment:
Are the regular methods defined in terms of the `-inline` ones?

Reply:
no, it is safe.